### ISSUE TITLE
Failed reproduction for ssd

### ISSUE THREAD

**Issue by Smith-xuan** [apple/ml-ssd#7]
Thank you for your work, but I encountered difficulties while trying to reproduce it.

I tried to reproduce your work on Qwen3-4B-Instruct. First, I filtered the seed partition of problems on RStar-Coder to obtain prompts. Since your dataset is not open-sourced, I applied your deduplication method and obtained over 30,000 entries. Then, I filtered based on the is_passed field in the dataset, keeping only entries where is_passed is true, which resulted in 10,167 entries — almost the same number as used in your work.

Next, I used Qwen3-4B-Instruct to sample this roughly 10k dataset once under the settings temp=1.6, topk=20, topp=0.8, and MAX_TOKENS=65536 (aligned with the paper) to generate the training set. Using this training set, I configured the following parameters:

{
    "global-batch-size": 32,
    "micro-batch-size": 1,
    "seq-length": 8192,
    "train-iters": 2500,
    "lr-warmup-iters": 250,
    "save-interval": 250,
    "lr": 5e-6,
    "min-lr": 1e-6,
    "lr-decay-style": "cosine",
    "adam-beta1": 0.9,
    "adam-beta2": 0.95,
    "weight-decay": 0.1
}

These parameters are fully aligned with the paper except for "seq-length": 8192 (the paper uses 64k, but statistics show that over 99th percentile of problems in the training set are under 8k).

Then, I tested on LiveCodeBench v5 and v6, and obtained the following results:
| Model | V5 pass@1 | V5 pass@5 | V6 pass@1 | V6 pass@5 |
|-------|-----------|-----------|-----------|-----------|
| Base Qwen3-4B-Instruct | 31.86% | 47.90% | 34.86% | 44.00% |
| SSD | 29.27% | 41.46% | 33.37% | 41.72% |
| Low-Temperature Self-Distillation | 29.76% | 43.90% | 33.37% | 42.94% |

It can be observed that the reproduced SSD not only did not improve performance, but actually showed a decline. The third experiment, which performed low-temperature distillation (temp=0.7) on the training set for comparison with SSD, also did not lead to any improvement and did not create a noticeable difference from SSD.

Could you tell me where the problem might have occurred?

**Comment by JianghaoWu** [comment]
Same issue here — the reproduced model performs worse than the base model (Qwen3-4B-Instruct). Could you consider releasing the training code and some data samples?

**Comment by Amanda2024** [comment]
Same issue here. But why is the response length so much shorter in our reproduced version compared to the [open-source checkpoint](https://huggingface.co/apple/SimpleSD-4B-instruct), which generates significantly longer outputs?

- ~49k with the [open-source checkpoint](https://huggingface.co/apple/SimpleSD-4B-instruct)
- ~2k with our reproduced version

**Comment by richardbaihe** [comment]
Thanks for your interests in reproducing our work! We're currently working with Apple's internal team for further code release. The data generation pipeline involves a few third-party libraries. We target to release them within the next 2 weeks and will update this thread as soon as it's available. Please stay tuned.

**Comment by richardbaihe** [comment]
Hi @Amanda2024 , @JianghaoWu  and @Smith-xuan , we just released our data generation pipeline here 15c429241729df2704f926bd3e6ac19cf502f245.

Sorry for any inconvenience and please feel free to reach out if you have any further questions. Thanks for all your interests!

**Comment by william-0g** [comment]
@richardbaihe 
Hi, thanks for releasing the data generation pipeline. I’m trying to reproduce the paper, but I’m confused whether this code matches the exact SSD setup. It seems to (1) generate on the full seed_sft train split instead of the deduplicated ~10K seed subset, and (2) not clearly support instruct vs. thinking/non-thinking generation with official chat templates. I also noticed the released prompt formatting and post-processing differ from the paper’s appendix. Could you clarify whether this is a simplified pipeline, or release the exact preprocessing, deduplication, prompt formatting, and generation settings used in the main experiments?

**Comment by richardbaihe** [comment]
Thanks for the question @william-0g . 

Yes, this is a simplified implementation adapted from our internal codebase, and we can verify that provided 4B-instruct example reproduces our paper's results.

**Comment by lukehansen** [comment]
I am also trying to reproduce the results -- @richardbaihe can you confirm if you used the `is_passed=True` filter in your dataset generation, the way OP suggests? That seems to be the only way to get ~10k questions, as reported in the paper.

**Comment by lukehansen** [comment]
@Smith-xuan for dataset generation the paper says they use 128K max tokens (the 64K is for training). I would be surprised if many of the Instruct samples exceed 64K, but could this explain why you cannot reproduce?

**Comment by richardbaihe** [comment]
> I am also trying to reproduce the results -- [@richardbaihe](https://github.com/richardbaihe) can you confirm if you used the `is_passed=True` filter in your dataset generation, the way OP suggests? That seems to be the only way to get ~10k questions, as reported in the paper.

Hey @lukehansen, you can enable is_passed=True to get the seed problem subset. 
We have some ablation on data quantity and we found 10k, 20k, 30k, ..., 80k all produce similar results which means the results are not sensitive to data amount (note, the minimal amount I tested is 10k, to reproduce our paper's results). 

Main suggestion is to follow our data generation pipeline in this repo which we have already verified that it reproduced our paper results. So even if you get more than 10k questions, it should be okay to reproduce.

**Comment by lukehansen** [comment]
Thank you @richardbaihe , that makes sense.

One more question: it seems even the Instruct model Qwen3-4B-Instruct-2507 will actually output CoT reasoning during the dataset generation ("I'll solve this step by step..."), and for some samples (<2%) it degenerates into long loops that would get cut off at the 65K token cap.

Did you filter these samples during SFT, or it is fine?

**Comment by richardbaihe** [comment]
> Thank you [@richardbaihe](https://github.com/richardbaihe) , that makes sense.
> 
> One more question: it seems even the Instruct model Qwen3-4B-Instruct-2507 will actually output CoT reasoning during the dataset generation ("I'll solve this step by step..."), and for some samples (<2%) it degenerates into long loops that would get cut off at the 65K token cap.
> 
> Did you filter these samples during SFT, or it is fine?

For this case, we didn't filter them out.

**Comment by lukehansen** [comment]
@richardbaihe Do you plan to release the generated dataset you used? Using the provided code and fine-tuning following the paper does not seem to reproduce the results.

**Comment by william-0g** [comment]
> [@richardbaihe](https://github.com/richardbaihe) Do you plan to release the generated dataset you used? Using the provided code and fine-tuning following the paper does not seem to reproduce the results.

The same.

**Comment by richardbaihe** [comment]
> [@richardbaihe](https://github.com/richardbaihe) Do you plan to release the generated dataset you used? Using the provided code and fine-tuning following the paper does not seem to reproduce the results.

Hi @lukehansen , we have constraints about releasing data generated from 3rd party model, however, we believe code for creating the data is better than the data itself. We have verified the generated data from this repo can reproduce 4B model results before we released the script. So we are quite sure about the data quality as long as you use our script to generate the data. 

If you cannot train a good model, try more frequent evaluation on your intermediate checkpoints to see if something is wrong. For instruct model, the results should be keep increasing until saturated after some steps (but according to my observation, Instruct model can very easily improved after tuned with the generated data, nearly all checkpoints and decoding params combinations can beat the baseline results. So given this observation, i would recommend to debug your training code with small amount of data. For thinking model, the best results comes earlier than the instruct model). 


For reference, here is 4b-instruct model tuned with this repo generated data:

<img width="762" height="554" alt="Image" src="https://github.com/user-attachments/assets/00e4b6d4-d2dc-4130-89ec-b7919180ddfb" />

**Comment by JeffLee1874** [comment]
Does ssd work for math reasoning? I try the same setting of qwen3-4b with deepmath-107k, but found performance dropped when evaluated on gsm8k, math-500.

> > [@richardbaihe](https://github.com/richardbaihe) Do you plan to release the generated dataset you used? Using the provided code and fine-tuning following the paper does not seem to reproduce the results.
> 
> Hi [@lukehansen](https://github.com/lukehansen) , we have constraints about releasing data generated from 3rd party model, however, we believe code for creating the data is better than the data itself. We have verified the generated data from this repo can reproduce 4B model results before we released the script. So we are quite sure about the data quality as long as you use our script to generate the data.
> 
> If you cannot train a good model, try more frequent evaluation on your intermediate checkpoints to see if something is wrong. For instruct model, the results should be keep increasing until saturated after some steps (but according to my observation, Instruct model can very easily improved after tuned with the generated data, nearly all checkpoints and decoding params combinations can beat the baseline results. So given this observation, i would recommend to debug your training code with small amount of data. For thinking model, the best results comes earlier than the instruct model).
> 
> For reference, here is 4b-instruct model tuned with this repo generated data:
> 
> <img alt="Image" width="762" height="554" src="https://private-user-images.githubusercontent.com/11585856/600337267-00e4b6d4-d2dc-4130-89ec-b7919180ddfb.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODA1NTc4MzUsIm5iZiI6MTc4MDU1NzUzNSwicGF0aCI6Ii8xMTU4NTg1Ni82MDAzMzcyNjctMDBlNGI2ZDQtZDJkYy00MTMwLTg5ZWMtYjc5MTkxODBkZGZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjA0VDA3MTg1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0MDU0NjZkNTdlZWJmYmQyYjRkYzZiOWUyZDA1MTcwOGYwZTcxZDYwZDNlMDY3YTk5MTFmYWMwYTExYzY0NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.W687PsZGvr9eSF2X-gcQ8rFb_ylM03UDsqNfpE_z07Y">
